### PR TITLE
Gem::Util.traverse_parents should not crash on permissions error

### LINKED ISSUE #2144: Util.traverse_parents fails with Errno::EACCES
Background:

I'm running a simple ruby program as a postfix filter. Postfix starts my program with clear environment and non-root user who does not have permissions to the current directory. I'm mentioning this because I have no-control over the environment my program is executed, it's enforced by Postfix.

This was working fine with rubygems 2.6, but it started crashing after upgrade to 2.7. Downgrade to 2.6 fixes the problem.

The issue initially occurred on ruby 2.4, but for testing purposes I will be using ruby 2.5. As far as I can say the behaviour does not change.

---

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [ ] The command line `gem`
  - [x] Other

---

Here are my current environment details:

```
$ gem env version
2.7.4
$ ruby -v
ruby 2.5.0p0 (2017-12-25 revision 61468) [x86_64-linux-musl]
```

---

Steps to reproduce the issue:

* create temporary working environment

```
mkdir -p /tmp/rubygems
cd /tmp/rubygems
```

* add `Dockerfile` with the following content:

```
FROM ruby:2.5-alpine

# Add non-root user and install random gem
RUN adduser -S example && gem install mail
# Our program, failing because BUNDLER_VERSION is not set
RUN echo "ENV['BUNDLER_VERSION'] = nil; require 'mail'; puts 'It works!';" > /bin/test.rb
# Set current directory (pwd) to a location where our user has no permissions
WORKDIR /root
USER example

CMD ["ruby", "/bin/test.rb"]
```

* build the image: `docker build -t rubygems-example .`
* run the container: `docker run --rm -it rubygems-example`
* you should see the error now

```
Traceback (most recent call last):
        16: from /bin/test.rb:1:in `<main>'
        15: from /usr/local/lib/ruby/site_ruby/2.5.0/rubygems/core_ext/kernel_require.rb:39:in `require'
        14: from /usr/local/lib/ruby/site_ruby/2.5.0/rubygems/core_ext/kernel_require.rb:128:in `rescue in require'
        13: from /usr/local/lib/ruby/site_ruby/2.5.0/rubygems.rb:213:in `try_activate'
        12: from /usr/local/lib/ruby/site_ruby/2.5.0/rubygems/specification.rb:1058:in `find_by_path'
        11: from /usr/local/lib/ruby/site_ruby/2.5.0/rubygems/specification.rb:1058:in `find'
        10: from /usr/local/lib/ruby/site_ruby/2.5.0/rubygems/specification.rb:1058:in `each'
         9: from /usr/local/lib/ruby/site_ruby/2.5.0/rubygems/specification.rb:1059:in `block in find_by_path'
         8: from /usr/local/lib/ruby/site_ruby/2.5.0/rubygems/bundler_version_finder.rb:46:in `compatible?'
         7: from /usr/local/lib/ruby/site_ruby/2.5.0/rubygems/bundler_version_finder.rb:12:in `bundler_version'
         6: from /usr/local/lib/ruby/site_ruby/2.5.0/rubygems/bundler_version_finder.rb:29:in `bundler_version_with_reason'
         5: from /usr/local/lib/ruby/site_ruby/2.5.0/rubygems/bundler_version_finder.rb:81:in `lockfile_version'
         4: from /usr/local/lib/ruby/site_ruby/2.5.0/rubygems/bundler_version_finder.rb:93:in `lockfile_contents'
         3: from /usr/local/lib/ruby/site_ruby/2.5.0/rubygems/util.rb:116:in `traverse_parents'
         2: from /usr/local/lib/ruby/site_ruby/2.5.0/rubygems/util.rb:116:in `loop'
         1: from /usr/local/lib/ruby/site_ruby/2.5.0/rubygems/util.rb:117:in `block in traverse_parents'
/usr/local/lib/ruby/site_ruby/2.5.0/rubygems/util.rb:117:in `chdir': Permission denied @ dir_chdir - /root (Errno::EACCES)
```

-------

To make the program work set `BUNDLER_VERSION` env to `1.16` (or something else).

The error appears to be coming from the `BundlerVersionFinder` that has been introduced in 2.7, but the root cause may actually be in `Gem::Util.traverse_parents` - perhaps it should rescue from `Errno::EACCES` and terminate the loop early instead of crashing?

---

Thanks very much for all the work you're putting into Bundler!

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### PR DESCRIPTION
**Pull request by robuye**
This change makes `traverse_parents` to skip directories it can't
read.

There are edge cases where program working directory (and parents)
is not "readable". If such a program includes `require 'some-gem'`
and environment variable BUNDLER_VERSION is not set then
`BundlerVersionFinder` will attempt to determine bundler version
by traversing all parent directories to find Gemfile.lock and determine
the bundler version from it. The program will then crash.

A real-world example is described here:

https://github.com/rubygems/rubygems/issues/2144

With this patch unreadable directories will be skipped and the
program won't crash.

Resolves #2144 

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by segiddins** [comment]
test seems to fail on windows: https://ci.appveyor.com/project/segiddins/rubygems/build/994/job/h5360tyk6t5wvg3h

**Comment by robuye** [comment]
Sorry about that, I have skipped the test on windows since `File.chmod` does not work. I did force push and the tests pass now.

The codeclimate complaint is fine I think? After all I have added complexity to the method and I don't think there is a way around it.

**Comment by segiddins** [comment]
👍 I don't know where the code climate check came from, anyways

File: test/rubygems/test_gem_util.rb [hunk 1]
```diff
@@ -29,6 +29,25 @@ def test_traverse_parents
     loop { break if enum.next.nil? } # exhaust the enumerator
   end
 
+  def test_traverse_parents_does_not_crash_on_permissions_error
+    skip 'skipped on MS Windows (chmod has no effect)' if win_platform?
+
+    FileUtils.mkdir_p 'd/e/f'
+    # remove 'execute' permission from "e" directory and make it
+    # impossible to cd into it and it's children
```
Comment (olleolleolle) [inline test/rubygems/test_gem_util.rb]: Grammar typo: `it's` => `its`

**Comment by robuye** [comment]
I did another force push correcting the typo in the comment. Thanks for careful review and guidance, I've been using rubygems for years and I'm very happy to be able to contribute back, even if it's a small edge case!

**Comment by robuye** [comment]
Looks like some unrelated tests on travis failed for ruby 1.8.7, would you guys like me to push again and force rebuild? Or is it something my change would cause and it's worth digging into?

**Comment by colby-swandale** [comment]
Thanks! @bundlerbot r+

**Comment by colby-swandale** [comment]
@bundlerbot retry

**Comment by hsbt** [comment]
@bundlerbot r+

**Comment by hsbt** [comment]
@bundlerbot retry
